crates/trie/db/src/commitment.rs

```diff
@@ -0,0 +1,38 @@
+use crate::{
+    DatabaseHashedCursorFactory, DatabaseProof, DatabaseStateRoot, DatabaseStorageRoot,
+    DatabaseTrieCursorFactory, DatabaseTrieWitness,
+};
+use reth_db::transaction::DbTx;
+use reth_trie::{
+    proof::Proof, witness::TrieWitness, KeccakKeyHasher, KeyHasher, StateRoot, StorageRoot,
+};
+
+/// The `StateCommitment` trait provides associated types for state commitment operations.
+pub trait StateCommitment: std::fmt::Debug + Send + Sync + Unpin + 'static {
+    /// The state root type.
+    type StateRoot<'a, TX: DbTx + 'a>: DatabaseStateRoot<'a, TX>;
+    /// The storage root type.
+    type StorageRoot<'a, TX: DbTx + 'a>: DatabaseStorageRoot<'a, TX>;
+    /// The state proof type.
+    type StateProof<'a, TX: DbTx + 'a>: DatabaseProof<'a, TX>;
+    /// The state witness type.
+    type StateWitness<'a, TX: DbTx + 'a>: DatabaseTrieWitness<'a, TX>;
```

Comment:
I'm a bit confused as to why these require `DbTx` bounds? Is it just because these must implement the `Database*` traits?

Reply:
This is a consequence of the impl of `DatabaseStateRoot` on `StateRoot`. As we see below we have a `DbTx` bound on `TX` which is required for certain methods used in the implementation. 
https://github.com/paradigmxyz/reth/blob/51594c9a68e437075f0aa0bfc13491dd82176d7c/crates/trie/db/src/state.rs#L132-L133

As a consequence of this in the implementation of `StateCommitment` for the `MerklePatriciaTrie` we must also have a `DbTx` bound on the `StateRoot` type as seen below:
```rust
impl StateCommitment for MerklePatriciaTrie {
    type StateRoot<'a, TX: DbTx + 'a> =
        StateRoot<DatabaseTrieCursorFactory<'a, TX>, DatabaseHashedCursorFactory<'a, TX>>;
        ...
}
```

Rust does not allow for an impl of a trait to have more strict bounds than the trait itself and therefore me must propagate the bound to the trait itself.  The above rationale is also applicable for the other types in `StateCommitment`.

Reply:
I just added this issue: https://github.com/paradigmxyz/reth/issues/12216

note that it should not block any of this work, but eventually these traits could be simplified